### PR TITLE
fix: 修复程序同时支持XEmbed和SNI时，托盘出现双图标问题

### PR DESCRIPTION
**Pull request by tsic404**
修复程序同时支持XEmbed和SNI时，托盘出现双图标问题

### CODE REVIEW

File: plugins/tray/trayplugin.cpp [hunk 1]
```diff
@@ -344,7 +344,11 @@ void TrayPlugin::sniItemsChanged()
         }
     }
 
+    m_registerSNIPID.clear();
+    QDBusConnection conn = QDBusConnection::connectToBus(QDBusConnection::SessionBus, "org.freedesktop.DBus");
```
Comment (zccrs) [inline plugins/tray/trayplugin.cpp:348]: direct use the QDBusConnection::sessionBus()

File: plugins/tray/xembedtraywidget.cpp [hunk 2]
```diff
@@ -574,3 +574,31 @@ bool XEmbedTrayWidget::isBadWindow()
...
+    status = XGetWindowProperty(display, winId, nameAtom, 0, 1024, 0,
+            XInternAtom(display, "CARDINAL", 0), &type, &format, &nitems, &after, &data);
+    if (status == Success && data) {
+        pid = *((long*)data);
+        XFree(data);
+        qDebug() << "_NET_WM_PID = " << pid;
+    }
+    else
+        qDebug() << "_NET_WM_PID not found";
+    return pid;
```
Comment (zccrs) [inline plugins/tray/xembedtraywidget.cpp:603]: Should close the X display if the "IS_WAYLAND_DISPLAY" is true

File: plugins/tray/trayplugin.cpp [hunk 3]
```diff
@@ -353,8 +357,15 @@ void TrayPlugin::xembedItemsChanged()
...
     QStringList trayKeyList;
+    QList<quint32> xembedOnlyWinIdList;
 
     for (auto winid : winidList) {
+        if (!m_registerSNIPID.contains(XEmbedTrayWidget::getWindowPID(winid))) {
+            xembedOnlyWinIdList.append(winid);
+        }
+    }
+
+    for (auto winid : xembedOnlyWinIdList) {
```
Comment (zccrs) [inline plugins/tray/trayplugin.cpp:368]: Can be processed together in the previous for-loop code block

File: plugins/tray/trayplugin.cpp [hunk 4]
```diff
@@ -344,7 +344,11 @@ void TrayPlugin::sniItemsChanged()
         }
     }
 
+    m_registerSNIPID.clear();
+    QDBusConnection conn = QDBusConnection::connectToBus(QDBusConnection::SessionBus, "org.freedesktop.DBus");
     for (int i = 0; i < sinTrayKeyList.size(); ++i) {
+        QString dbusName = SNITrayWidget::serviceAndPath(itemServicePaths.at(i)).first;
+        m_registerSNIPID.insert(conn.interface()->servicePid(dbusName));
         traySNIAdded(sinTrayKeyList.at(i), itemServicePaths.at(i));
```
Comment (zccrs) [inline plugins/tray/trayplugin.cpp:352]: Check the PID also for the SNI item

**Review by zccrs — CHANGES_REQUESTED** [review]

File: plugins/tray/abstracttraywidget.h [hunk 5]
```diff
@@ -59,6 +60,8 @@ class AbstractTrayWidget: public QWidget
     const QRect perfectIconRect() const;
     void resizeEvent(QResizeEvent *event) override;
 
+    uint m_ownerPID;
```
Comment (zccrs) [inline plugins/tray/abstracttraywidget.h:63]: Move the member to the private scope

File: plugins/tray/trayplugin.cpp [hunk 6]
```diff
@@ -327,13 +327,15 @@ void TrayPlugin::initSNI()
 void TrayPlugin::sniItemsChanged()
 {
     const QStringList &itemServicePaths = m_sniWatcher->registeredStatusNotifierItems();
+    QDBusConnection conn = QDBusConnection::sessionBus();
```
Comment (zccrs) [inline plugins/tray/trayplugin.cpp:330]: The "conn" is first used at the 350 line, should define it at the first use nearby.

File: plugins/tray/trayplugin.cpp [hunk 7]
```diff
@@ -343,40 +345,44 @@ void TrayPlugin::sniItemsChanged()
...
 }
 
 void TrayPlugin::xembedItemsChanged()
 {
     QList<quint32> winidList = m_trayInter->trayIcons();
-    QStringList trayKeyList;
-    QList<quint32> xembedOnlyWinIdList;
+    QStringList newlyAddedTrayKeyList;
+    QStringList allKeytary;
+    QList<quint32> newlyAdded;
```
Comment (zccrs) [inline plugins/tray/trayplugin.cpp:364]: newlyAddedWindowID

File: plugins/tray/trayplugin.cpp [hunk 8]
```diff
@@ -343,40 +345,44 @@ void TrayPlugin::sniItemsChanged()
...
         }
     }
-
-    m_registerSNIPID.clear();
-    QDBusConnection conn = QDBusConnection::connectToBus(QDBusConnection::SessionBus, "org.freedesktop.DBus");
     for (int i = 0; i < sinTrayKeyList.size(); ++i) {
         QString dbusName = SNITrayWidget::serviceAndPath(itemServicePaths.at(i)).first;
-        m_registerSNIPID.insert(conn.interface()->servicePid(dbusName));
-        traySNIAdded(sinTrayKeyList.at(i), itemServicePaths.at(i));
+        uint pid = conn.interface()->servicePid(dbusName);
```
Comment (zccrs) [inline plugins/tray/trayplugin.cpp:350]: Like the XEmbedTrayWidget::getWindowPID, should get the pid of the dbus item at there by the SNITrayWidget::xxxx.

**Review by zccrs — CHANGES_REQUESTED** [review]

**Comment by YihaoPeng** [comment]
支持。正在被该问题困扰。

image

**Comment by tsic404** [comment]
Internally merged and released version 5.4.69-1

**Comment by YihaoPeng** [comment]
It looks like the update hasn't reached UOS home edition yet.

image

Regardless of the technical details, as a user, I just want a well-rendered tray area instead of having 4 duplicated icons.

**Comment by zccrs** [comment]
> 看起来这个更新还没有到达UOS家庭版。
> 
> image
> 
> 作为用户，我只想有一个正常显示的托盘区域，而不是像这样出现4个相同的图标。

家庭版的更新周期和策略跟deepin不一致，你可以把问题报告到家庭版，这样才有希望尽快推送更新解决。

**Pull request closed without merging**